File: src/main/java/net/hectus/neobb/turn/card_game/InteractableCardTurn.java
package net.hectus.neobb.turn.card_game;

import net.hectus.neobb.player.NeoPlayer;
import org.bukkit.block.Block;

public abstract class InteractableCardTurn extends CardTurn {
    private int interactions = 0;
    private long lastInteraction = System.currentTimeMillis();

    public InteractableCardTurn(NeoPlayer player) { super(player); }
    public InteractableCardTurn(Block data, NeoPlayer player) { super(data, player); }

    public int maxInteractions() { return 3; }
    public long interactionIntervalMs() { return 1000; }

    public synchronized void interact() {
        if (interactions <= maxInteractions() && System.currentTimeMillis() - lastInteraction <= interactionIntervalMs()) {
            lastInteraction = System.currentTimeMillis();
            interactions++;
            player.nextPlayer().damage(damage());
        }
    }
}

File: src/main/kotlin/net/hectus/neobb/modes/turn/card_game/InteractableCardTurn.kt
package net.hectus.neobb.modes.turn.card_game

import com.marcpg.libpg.storing.Cord
import net.hectus.neobb.player.NeoPlayer
import org.bukkit.block.Block

abstract class InteractableCardTurn(data: Block?, cord: Cord?, player: NeoPlayer?) : CardTurn(data, cord, player) {
    var interactions: Int = 0
        private set
    var lastInteraction: Long = System.currentTimeMillis()
        private set

    open val maxInteractions = 3
    open val interactionIntervalMs = 1000

    @Synchronized
    fun interact() {
        if (interactions <= maxInteractions && System.currentTimeMillis() - lastInteraction <= interactionIntervalMs) {
            lastInteraction = System.currentTimeMillis()
            interactions++
            player?.nextPlayer()?.damage(damage)
        }
    }
}
